tests/specs/lemmas.k

```diff
@@ -75,25 +65,29 @@ module LEMMAS
   // Buffer Reasoning
   // ########################
 
-    rule WS ++ .ByteArray => WS  [simplification]
-    rule .ByteArray ++ WS => WS  [simplification]
+    rule WS         ++ .ByteArray => WS  [simplification]
+    rule .ByteArray ++ WS         => WS  [simplification]
 
   // ########################
   // Map Reasoning
   // ########################
 
-    // re-order assignments
-    rule MEM [ K1 := BUF1:ByteArray ] [ K2 := BUF2:ByteArray ] => MEM [ K2 := BUF2 ] [ K1 := BUF1 ]  requires K1  >Int K2 andBool K1 -Int K2 >=Int #sizeByteArray(BUF2)                           [simplification]
+    rule #lookup ( _M:Map [ K1 <- V1 ] , K2 ) => #lookup ( K1 |-> V1 , K1 )  requires K1 ==Int  K2 [simplification]
+    rule #lookup (  M:Map [ K1 <- _  ] , K2 ) => #lookup ( M         , K2 )  requires K1 =/=Int K2 [simplification]
 
-    // overwritten assignment
-    rule MEM [ K1 := BUF1           ] [ K2 := BUF2           ] => MEM [ K2 := BUF2 ]                 requires K2 <=Int K1 andBool K1 +Int #sizeByteArray(BUF1) <=Int K2 +Int #sizeByteArray(BUF2) [simplification]
+    rule 0 <=Int #lookup( _M:Map , _ )             => true [simplification, smt-lemma]
+    rule         #lookup( _M:Map , _ ) <Int pow256 => true [simplification, smt-lemma]
 
+    // re-order assignments
+    rule MEM [ K1 := BUF1:ByteArray ] [ K2 := BUF2:ByteArray ] => MEM [ K2 := BUF2 ] [ K1 := BUF1 ]  requires         #sizeByteArray(BUF2) <=Int K1 -Int K2                   andBool K2  <Int K1 [simplification]
+    // overwritten assignment
+    rule MEM [ K1 := BUF1           ] [ K2 := BUF2           ] => MEM [ K2 := BUF2 ]                 requires K1 +Int #sizeByteArray(BUF1) <=Int K2 +Int #sizeByteArray(BUF2) andBool K2 <=Int K1 [simplification]
     // redundant assignment
-    rule MEM [ K  := BUF1           ] [ K  := BUF2           ] => MEM [ K  := BUF2 ]                 requires #sizeByteArray(BUF1) ==Int #sizeByteArray(BUF2)                                     [simplification]
+    rule MEM [ K  := BUF1           ] [ K  := BUF2           ] => MEM [ K  := BUF2 ]                 requires         #sizeByteArray(BUF1) ==Int #sizeByteArray(BUF2)                             [simplification]
```

Comment:
Here's an example where the second works but not the first:
```MEM [ 2 := "1" ] [ 1 := "123" ]```

And the first doesn't look like 'overwritten' assignment, the two assignments are disjoint and are just being reordered.

Reply:
Oh, I realized you may have been referring to the last two of the three and not the first two...

Reply:
I also think @rv-jenkins (aka @ehildenb) was probably referring to the last two.

I think you're right, if K1 ==Int K2 then the second rule implies the third rule, since then
- the first side condition of the second rule simplifies to the first side condition of the third rule:
```
K1 +Int #sizeByteArray(BUF1) <=Int K2 +Int #sizeByteArray(BUF2)
=> (since K1 ==Int K2)
K1 +Int #sizeByteArray(BUF1) <=Int K1 +Int #sizeByteArray(BUF2)
=>
#sizeByteArray(BUF1) <=Int #sizeByteArray(BUF2)
```
- the second side condition of the second rule simplifies to true (and the third rule has no second side condition)
```
K2 <=Int K1
=> (since K1 ==Int K2)
true
```

I removed the redundant rule here: https://github.com/kframework/evm-semantics/pull/1081/commits/c8b4f79e32e2fdb99a163445648633e40a69a1e8